### DISCUSSION TITLE
running code on MacOS / Jupyter Notebook files / Gym OpenGL on MacOS

### DISCUSSION THREAD

**Discussion by bayareacoder** [PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook#18] [General]
It seems your 'Getting Started' is only for Linux. Do you have instructions for MacOS? In particular how to install/use OpenGL. I have this when running code of your 1st example in Ch1 in Jupyter Notebook:
```
ImportError: Can't find framework /System/Library/Frameworks/OpenGL.framework
```
It would be really nice to have your source files in Jupyter Notebook format (.pynb). 
Much easier to follow along as the examples execute.
A good example of an easy-to-use repo is this:
https://github.com/stefan-jansen/machine-learning-for-trading

UPDATE: I figured this out. You need to update "pyglet" package in the "gym" install to the latest version for rendering to work on MacOS, per:https://github.com/openai/gym/issues/2101
Update it in Anaconda via CLI: `conda install -c conda-forge pyglet`
Rendering works for me now after pyglet was updated to v1.5.16.

**Comment by praveen-palanisamy** [comment]
Hi @bayareacoder,

Thanks for reporting your issue and following it up with a resolution. I am glad that updated `pyglet` solved your issue on Mac OSX.
The pyglet version and the gym versions have dependencies so I need to check which pair works and I will update the [conda env yaml file](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/master/tfrl-cookbook.yml) after testing.

**Comment by praveen-palanisamy** [comment]
Also, I have created issue # 14 to track your useful suggestion on adding Jupyter notebooks for the script. I will get them added by this weekend.

**Comment by bayareacoder** [comment]
The issue I reported is only one. Still not able to install all packages as described in your setup guide from 'tfrl-cookbook.yml' and as a result, probably due to incompatible versions that come with the Anaconda install, many scripts from your book don't seem to work.  I would strongly suggest you test your setup instructions, and scripts, on MacOS to avoid frustration with readers on MacOS. It seems a good and practical book but if the code doesn't run obviously it's an issue.

```
conda env create -f tfrl-cookbook.yml -n "tfrl-cookbook" 
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - libxcb=1.14
  - libgcc-ng=9.1.0
  - cudatoolkit=10.1.243
  - libstdcxx-ng=9.1.0
  - cudnn=7.6.5
```

Here is an example of run-time error in Ch5 (no static error on import, so all packages seem present):
```
python stock_trading_visual_env.py
Traceback (most recent call last):
  File "stock_trading_visual_env.py", line 195, in <module>
    obs = env.reset()
  File "stock_trading_visual_env.py", line 100, in reset
    return self.get_observation()
  File "stock_trading_visual_env.py", line 125, in get_observation
    img_observation = self.viz.render_image_observation(
  File ".../Chapter05/trading_utils.py", line 132, in render_image_observation
    fig_data = np.frombuffer(self.fig.canvas.tostring_rgb(), dtype=np.uint8)
  File "/Users/bart/opt/anaconda3/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py", line 431, in tostring_rgb
    return self.renderer.tostring_rgb()
AttributeError: 'FigureCanvasMac' object has no attribute 'renderer'
```

**Comment by bayareacoder** [comment]
> Also, I have created issue # 14 to track your useful suggestion on adding Jupyter notebooks for the script. I will get them added by this weekend.

Looking forward to an update - thanks.

**Comment by praveen-palanisamy** [comment]
Hi @bayareacoder ,
Thanks for the ping! I'm still generating and testing the notebooks but I should be able to provide you the notebooks for the recipes in some of the chapters today/tomorrow that you can use while I am preparing the others. I'll update you when the initial set of notebooks are ready.

**Comment by shruthis-shetty** [comment]
Hi @bayareacoder ,
Have you got solution to this issue?

**Comment by bayareacoder** [comment]
> Hi @bayareacoder ,
> 
> Have you got solution to this issue?

Nope. Still waiting on the author to fix so it works on MacOS.

**Comment by shruthis-shetty** [comment]
> > Hi @bayareacoder ,
> > Have you got solution to this issue?
> 
> Nope. Still waiting on the author to fix so it works on MacOS.

Cool. We'll contact the author and try to provide the solution at the earliest. Thanks.

**Comment by praveen-palanisamy** [comment]
@bayareacoder , I have updated the repo with Jupyter/IPython notebooks files for each of the recipes in [Chapter 1](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/tree/master/Chapter01) [Chapter 2](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/tree/master/Chapter02) and [Chapter 3](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/tree/master/Chapter03). The remaining will be added when they are ready.
Pull the latest code, check them out and let me know if you run into any issues with them.

**Comment by praveen-palanisamy** [comment]
@bayareacoder : You have mentioned a few issues on Mac OSX in your previous messages. I'm going to convert this issue thread into a discussion. Please create separate issues with description for each specific item to track/fix.

**Reply by praveen-palanisamy** [reply]
I have created another issue out of this thread: #19

**Comment by bayareacoder** [comment]
As reported here: https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/issues/14#issuecomment-844380015, I'm able to run the .ipynb files for Ch1-3. I plan to focus on the .ipynb instead of .py files going forward.

For the Getting Started to work on MacOS, besides the issue mentioned at the top of this thread, you may want to mention that the system tools for MacOS can be installed via HomeBrew: `brew install make cmake ffmpeg`

**Reply by praveen-palanisamy** [reply]
Thanks for the suggestion! I have updated the [Getting started](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook#getting-started) section in the README with the brew install command.
